[jeff] iP

jefrai opened the pull request:
# Project Miki 🎀✨

> The most merciful thing in the world, I think, is the inability of the human mind to correlate all its contents.
> <p align="right">-ef1500</p>

> Ogey rrat
> _<p align="right">P. Usada et al.</p>_

**Miki frees your mind from its physical locus.**
Miki is
- text-based
- [here](https://github.com/jefrai/ip/releases "there")
- transient
- eternal

**All you need to do is:**

1. download from [here](https://github.com/jefrai/ip/releases "miki/releases")!
2. `java -jar miki.jar`
3. add your tasks!
4. ???!
5. profit!

---

### With Project Miki, you can:

- [X] Plan your days with Miki!
- [ ] Autorun arbitrary files with Miki! (coming ~~soon~~ (time is subjective))
- [X] Get berated by Miki! (for poor spelling)
- [X] Procrastinate on your deadlines with Miki!
- [ ] save her

```java
    @Override
    public void run(TaskList tasks, Ui ui, Storage storage) {
        tasks.add(new Deadline("", LocalDateTime.of(2021, 05, 21, 15, 56, 25)));
        tasks.get(tasks.size() - 1).mark();
    }
```

> ashes to ashes, dust to dust
> _<p align="right">M. Ueda</p>_

Commit:
Add Gradle support

jefrai committed:
Add echoing, name project

jefrai committed:
Replace echoing with list-keeping

jefrai committed:
Add task marking

jefrai committed:
Add specialised task types

jefrai committed:
Add batch CLI testing

jefrai committed:
Add input error handling, remove adding typeless tasks

jefrai committed:
Add task deletion, improve error handling

jefrai committed:
Add filesystem saving/loading

jefrai committed:
Add date & time support, enforce task timings

jefrai committed:
Merge branch 'branch-Level-7'

jefrai committed:
Merge branch 'branch-Level-8'

jefrai committed:
Refactor sources to fit OOP

jefrai committed:
Merge branch 'add-gradle-support'

jefrai committed:
Add JUnit tests, fix gradle support

jefrai committed:
Add JavaDoc comments

jefrai committed:
Refactor sources to fit coding standard

jefrai committed:
Add task finding

jefrai committed:
Merge branch 'branch-A-JavaDoc'

jefrai committed:
Merge branch 'branch-A-CodingStandard'

jefrai committed:
Merge branch 'branch-Level-9'

ExtraShotLatte commented on src/main/java/Miki.java:
```diff
@@ -0,0 +1,59 @@
+import command.Command;
+import shigure.Ui;
+import storage.Storage;
+import task.Parser;
+import task.TaskList;
+
+import java.io.IOException;
```
Great ordering of import statements!

ExtraShotLatte commented on src/main/java/Miki.java:
```diff
@@ -0,0 +1,59 @@
+            try {
+                storage.load("autosave.txt", tasks);
+            } catch (IOException | Storage.MikiLoadException ex) {
+
```
Would it be better if the user can see that the loading has error?

jefrai replied:
True - no message is raised if there's no autosave file since that's 'normal' when we're running for the first time,
but now if the file exists but is corrupt there's a new message to alert the user.

ExtraShotLatte commented on src/main/java/Miki.java:
```diff
@@ -0,0 +1,59 @@
+    public static void main(String[] args) {
+        boolean hasAsciiOnly = false;
+        boolean hasNoAutoload = false;
+        for (int i = 0; i < args.length; i++) {
+            if (args[i].equals("--ascii-only")) {
+                hasAsciiOnly = true;
+            }
+            if (args[i].equals("--no-autoload")) {
+                hasNoAutoload = true;
```
I really like the feature of no autoloading of task!

ExtraShotLatte commented on src/main/java/command/Find.java:
```diff
@@ -0,0 +1,29 @@
+    @Override
+    public void run(TaskList tasks, Ui ui, Storage storage) {
+        int matches = 0;
+        for (int i = 0; i < tasks.size(); i++) {
+            if (tasks.get(i).hasMatchingObjective(regex)) {
+                matches++;
+            }
+        }
+        ui.print("here's your " + matches + (matches == 1 ? " match:" : " matches:"));
+        for (int i = 0; i < tasks.size(); i++) {
+            if (tasks.get(i).hasMatchingObjective(regex)) {
+                ui.print(i + 1 + ". " + tasks.get(i));
```
Can we run the linear search only once?

jefrai replied:
Yep, probably a good idea to keep the expensive regexes to a minimum - adjusted accordingly!

ExtraShotLatte commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,147 @@
+                }
+            }
+        } catch (NumberFormatException | TaskParseException ex) {
+            throw new MikiLoadException("this file is corrupt...");
```
Corrupted instead of corrupt?

jefrai replied:
👀 Both are grammatical actually, but looks like corrupted is a lot more common when referring to files so switched to that

ExtraShotLatte commented on src/main/java/task/Deadline.java:
```diff
@@ -0,0 +1,139 @@
+    public static Deadline parseArgs(String[] args) throws TaskParseException {
+        String objective = "";
+        String by = "";
+        boolean isInTokenBy = false;
```
Perharps can explain what is the situation that this boolean is used for.

jefrai replied:
Added a little comment to that line & similar lines elsewhere 👍

ExtraShotLatte commented on src/main/java/task/Deadline.java:
```diff
@@ -0,0 +1,139 @@
+        return repres.toArray(new String[repres.size()]);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
```
Great use of inheritDoc in here and other parts of the code!

ExtraShotLatte reviewed (commented):
Great personality 👍

LGTM! Great job in following the coding conventions as well!

jefrai committed:
Add PR Review suggested improvements

jefrai committed:
Add Checkstyle tests

leonpoh reviewed (commented):
Looks good a lot of abstraction following single responsibility principle

jefrai committed:
Add GUI

jefrai committed:
Merge branch 'branch-Level-10'

jefrai committed:
Fix JAR generation

The previous commit caused non-functional Java Archive files to be
created as the main() method was placed in a class extending JavaFX's
Application.

JAR generation needs to be restored to allow for easier usage of the
project.

Let's rectify JAR generation by moving the main() method into a
different class.

jefrai committed:
Add Assertions

There are portions of code which expect arguments to be non-null, or
call methods which are expected to have certain behaviours.

If these code elements encounter incorrect arguments or method
behaviour, they may throw difficult-to-diagnose Exceptions or fail
silently.

Let's add assertions to these portions of code to make it easier to
diagnose such situations, if they arise due to bugs elsewhere in this
project or in external Java modules.

Using assertions allows a visible and clear error to be thrown when an
unexpected situation arises, with the affected source line indicated and
a relevant message of choice included.

jefrai committed:
Refactor sources to code quality guidelines

There are sections of code which deviate from code quality guidelines.

This makes the code harder to read and understand, and thus more
difficult to debug.

Let's,
* improve code formatting, spacing and variable naming
* add better comments to less-obvious code actions
* perform method extraction on segments with deep nesting

jefrai committed:
Merge pull request #1 from jefrai/branch-A-Assertions

Add Assertions

jefrai committed:
Merge branch 'master' into branch-A-CodeQuality

jefrai committed:
Merge pull request #2 from jefrai/branch-A-CodeQuality

Refactor sources to code quality guidelines

jefrai committed:
Add natural language date support

jefrai committed:
Merge pull request #4 from jefrai/branch-C-NaturalDates

Add natural language date support

jefrai committed:
Add User Guide

jefrai committed:
Fix Checkstyle violations

jefrai committed:
Improve natural language date parsing

jefrai committed:
Improve code quality naming
